pwnlib/encoders/i386/ascii_shellcode.py

```diff
@@ -0,0 +1,300 @@
+
+class AsciiShellcodeEncoder(Encoder):
+    """ Ascii encoder based on:
+    http://julianor.tripod.com/bc/bypass-msb.txt
+    
+    A more visual explanation:
+    https://github.com/VincentDary/PolyAsciiShellGen/blob/master/README.md#mechanism
+
+    See the docstring of `__init__` and `__call__`
+    """
+
+    def __init__(self, slop = 20):
+        """ Init
+
+        Args:
+            slop (int): The amount esp will be increased by in the allocation
+            phase (In addition to the length of the packed shellcode) as well
+            as defines the size of the NOP sled (you can increase/decrease the
+            size of the NOP sled by adding/removing b'P'-s to/from the end of
+            the packed shellcode)
+        """
+        if six.PY2:
+            super(AsciiShellcodeEncoder, self).__init__()
+        elif six.PY3:
+            super().__init__()
+        self.slop = slop
+
+    # def asciify_shellcode(shellcode, slop, vocab = None):
+    @LocalContext
+    def __call__(self, raw_bytes, avoid=None, pcreg=None):
+        r""" Pack shellcode into only ascii characters that unpacks itself and
+        executes (on the stack)
+
+        Args:
+            raw_bytes (bytes): The shellcode to be packed
+            avoid (set, optional): Characters to avoid. Defaults to allow
+            printable ascii (0x21-0x7e).
+            pcreg (NoneType, optional): Ignored
+
+        Raises:
+            RuntimeError: A required character is not in ``vocab``
+            RuntimeError: Not supported architecture
+            ArithmeticError: The allowed character set does not contain
+            two characters that when they are bitwise-anded with eachother
+            they result is 0
+
+        Returns:
+            bytes: The packed shellcode
+
+        Examples:
+
+            >>> context.update(arch='i386', os='linux')
+            >>> sc = b"\x83\xc4\x181\xc01\xdb\xb0\x06\xcd\x80Sh/ttyh/dev\x89\xe31\xc9f\xb9\x12'\xb0\x05\xcd\x80j\x17X1\xdb\xcd\x80j.XS\xcd\x801\xc0Ph//shh/bin\x89\xe3PS\x89\xe1\x99\xb0\x0b\xcd\x80"
+            >>> encoders.i386.ascii_shellcode.encode(sc)
+            b'TX-!!!!-"_``-~~~~P\\%!!!!%@@@@-!6!!-V~!!-~~<-P-!mha-a~~~P-!!L`-a^~~-~~~~P-!!if-9`~~P-!!!!-aOaf-~~~~P-!&!<-!~`~--~~~P-!!!!-!!H^-+A~~P-U!![-~A1~P-,<V!-~~~!-~~~GP-!2!8-j~O~P-!]!!-!~!r-y~w~P-c!!!-~<(+P-N!_W-~1~~P-!!]!-Mn~!-~~~<P-!<!!-r~!P-~~x~P-fe!$-~~S~-~~~~P-!!\'$-%z~~P-A!!!-~!#!-~*~=P-!7!!-T~!!-~~E^PPPPPPPPPPPPPPPPPPPPP'
+            >>> sc = shellcraft.echo("Hello world") + shellcraft.exit()
+            >>> ascii = encoders.i386.ascii_shellcode.encode(asm(sc))
+            >>> ascii += asm('jmp esp')
+            >>> ELF.from_bytes(ascii).process().recvall()
+            b'Hello world'
+        """
+        if not avoid:
+            vocab = b"!\"#$%&'()*+,-./0123456789:;<=>?@ABCDEFGHIJKLMNOPQRSTUVWXYZ[\\]^_`abcdefghijklmnopqrstuvwxyz{|}~"
+        else:
+            required_chars = set(b'\\-%TXP')
+            allowed = set(map(ord, all_chars))

```

Comment:
Is `all_chars` really a string? It should be a bytearray, so you could just do a `set(all_chars)`... nevermind

Reply:
I'm entirly sure what you mean, but I think I fixed it here https://github.com/Gallopsled/pwntools/pull/1667/commits/d685b4174d1dd5e3db28f4d3b4a7f9d2021e5eb2.